certora/specs/CER-11-Controllers/CER-12-Controllers-number.spec

```diff
@@ -7,20 +7,63 @@ Each Account can have at most one Controller Vault enabled at a time unless it's
 
 methods {
     function numOfController(address account) external returns(uint8) envfree;
+    function getExecutionContextAreChecksDeferred() external returns(bool) envfree;
+    function containsStatusCheckFor(address account) external returns(bool) envfree;
 }
 
 /**
  * Check that we never have more than one controller.
  *
- * It does not work for batch(), as it requires rather complex reasoning. Some
+ * It does not work for batch() or call(), as it requires rather complex reasoning. Some
  * pieces necessary for this reasoning:
  * - the account status checks verifies that there is only a single controller.
  * - the only place that adds a controller issues an account status check for
  *   the respective account, either immediately or by registering it to the set.
- * - at the end of batch(), the execution context is restored.
+ * - whenever the setDeferredChecks bit in `executionContext` is cleared,
+ *   we make sure that we run the account status checks for
+ *   all accounts registered for a status check.
+ * 
+ * We prove this separately for enableController as we need to break it into
+ * cases (when checks are deferred vs not). This is done in the 3 rules that
+ * follow this one.
  */
 invariant onlyOneController(address a)
     numOfController(a) <= 1
     filtered { // TODO: make this work for batch as well
         f -> f.selector != sig:batch(IEVC.BatchItem[] calldata).selector
+            && f.selector != sig:call(address, address, uint256, bytes calldata).selector
+            && f.selector != sig:enableController(address, address).selector
     }
+
+// For enableController, we can only check this invariant directly if 
+// account status checks are not deferred. To cover the deferred case,
+// we check that when account checks are deferred, the enableController
+// call will enqueue a check for the relevant account using the rule
+// `enableControllerEnqueuesStatusCheckWhenDeferred`
+invariant onlyOneControllerEnableController(address a)
+    numOfController(a) <= 1 && !getExecutionContextAreChecksDeferred()
+    filtered {
+        f -> f.selector == sig:enableController(address, address).selector
+    }
```

Comment:
so OnlyController is not an invariant, right? it can break due to  enableController
and then what happens when numOfController(a) >= 2 on the other functions, should they all revert?

Reply:
if checks are deferred (transient state within a checks-deferrable call), we can have more than one controller at the time. in such a case, having at most one controller should be enforced at the time of the account status check.

if checks are not deferred, we can have at most one controller. any time `enableController` is called (the only function that enables controllers), it should revert if enabling new controller would result in having more than one at the time

Reply:
I rewrote these as rules instead of invariants to address @nd-certora 's point. (The rules should work the same as the previous invariants but making them rules makes it clearer how they work)